src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/NestedTypesShouldNotBeVisible.cs

```diff
@@ -102,6 +103,14 @@ public override void Initialize(AnalysisContext analysisContext)
                                 return;
                             }
 
+                            // Allow builder pattern
+                            if (nestedType.Name.EndsWith("Builder", StringComparison.Ordinal) &&
+                                containingType.ContainingType == null &&
+                                !containingType.InstanceConstructors.Any(ctor => ctor.IsExternallyVisible()))
```

Comment:
Do you want to have other constraints? Do you want to keep the depth constraint I have added?

Reply:
I think this looks fine.